openapi-cli/src/main/java/org/ballerinalang/ballerina/service/OpenApiResourceMapper.java

```diff
@@ -213,34 +246,230 @@ private void addResourceParameters(FunctionDefinitionNode resource, OperationAda
+                        if (recordVariable.typeDescriptor() instanceof TypeReferenceTypeSymbol) {
+                            TypeReferenceTypeSymbol typeRecord =
+                                    (TypeReferenceTypeSymbol) recordVariable.typeDescriptor();
+                            handleRecordPayload(queryParam, definitions, typeRecord);
+                        }
+                    }
+                    if (property instanceof ArrayProperty) {
+                        setArrayProperty(queryParam, definitions, field, (ArrayProperty) property);
+                    }
+                    messageModel.property(field.name(), property);
+                }
+            }
+        }
+
+        if (!definitions.containsKey(componentName)) {
+            //Set properties for the schema
+            definitions.put(componentName, messageModel);
+            this.openApiDefinition.setDefinitions(definitions);
+        }
+    }
+
+    private void setArrayProperty(RequiredParameterNode queryParam, Map<String, Model> definitions, FieldSymbol field,
+                                  ArrayProperty property) {
+
+        TypeSymbol symbol = field.typeDescriptor();
+        int arrayCount = 0;
+        while (symbol instanceof ArrayTypeSymbol) {
+            arrayCount = arrayCount + 1;
+            ArrayTypeSymbol arrayTypeSymbol = (ArrayTypeSymbol) symbol;
+            symbol = arrayTypeSymbol.memberTypeDescriptor();
+        }
+        //handle record field has nested record array type ex: Tag[] tags
+        Property symbolProperty  = getOpenApiProperty(symbol.typeKind().getName());
+        if (symbolProperty instanceof RefProperty) {
+            RefModel refModel = new RefModel();
+            refModel.setReference(symbol.name().trim());
+            ((RefProperty) symbolProperty).set$ref(refModel.get$ref());
+
+            Optional<Symbol> recordSymbol = semanticModel.symbol(queryParam.syntaxTree().filePath(),
+                    LinePosition.from(field.location().lineRange().startLine().line(),
+                            field.location().lineRange().startLine().offset()));
+            //Set the record model to the definition
+            VariableSymbol recordVariable = (VariableSymbol) recordSymbol.orElseThrow();
+            ArrayTypeSymbol arrayTypeSymbol = (ArrayTypeSymbol) recordVariable.typeDescriptor();
+            if (arrayTypeSymbol.memberTypeDescriptor() instanceof TypeReferenceTypeSymbol) {
+                TypeReferenceTypeSymbol typeRecord =
+                        (TypeReferenceTypeSymbol) arrayTypeSymbol.memberTypeDescriptor();
+                handleRecordPayload(queryParam, definitions, typeRecord);
+            }
+        }
+        //Handle nested array type
+        if (arrayCount > 1) {
+            ArrayProperty input = new ArrayProperty();
+            input.setItems(new ArrayProperty());
+            property.setItems(handleArray(arrayCount - 1, symbolProperty, input));
```

Comment:
Can we try to move the creation of the first array inside the function as well ?